Optimize get_json_object Spark function using simdjson

PHILO-HE opened the pull request:
This PR proposes an implementation for Spark `get_json_object` function based on simdjson lib. This function returns a json object, represented by VARCHAR, from json string by searching user-specified path.

Spark source code [link](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala#L119).

rui-mo reviewed (commented):
Can we provide some background for this PR? Please also include Spark's implementation in the PR description.

rui-mo commented on velox/functions/sparksql/tests/JsonTest.cpp:
```diff
@@ -0,0 +1,105 @@
+}
+
+} // namespace
+} // namespace facebook::velox::functions::sparksql::test
```
Add empty line at the end.

wanweiqiangintel commented:
> _ No description provided. _
@PHILO-HE, Could you please add some description to make the PR clear?

PHILO-HE commented:
@rui-mo, please take a review. Thanks!

rui-mo commented on velox/functions/sparksql/SIMDJsonFunctions.h:
```diff
@@ -0,0 +1,192 @@
+  FOLLY_ALWAYS_INLINE std::string getFormattedJsonPath(
+      const arg_type<Varchar>& jsonPath) {
+    // Makes a conversion from spark's json path, e.g.,
+    // converts "$.a.b" to "/a/b".
```
Can we move this comment above `getFormattedJsonPath`, and list all the rules for conversion? E.g. '$' will be ignored, '[' -> '/' etc.

rui-mo commented on velox/functions/sparksql/SIMDJsonFunctions.h:
```diff
@@ -0,0 +1,192 @@
+        }
+      }
+      case ondemand::json_type::boolean: {
+        bool boolResult = false;
```
The init value seems to be not needed.

rui-mo commented on velox/functions/sparksql/SIMDJsonFunctions.h:
```diff
@@ -0,0 +1,192 @@
+          result.append("true");
+        } else {
+          result.append("false");
+        }
```
result.append(boolResult ? "true" : "false");

rui-mo commented on velox/functions/sparksql/SIMDJsonFunctions.h:
```diff
@@ -0,0 +1,192 @@
+      }
+      case ondemand::json_type::boolean: {
+        bool boolResult = false;
+        rawResult.get_bool().get(boolResult);
```
Do we need to acquire the error code?

rui-mo commented on velox/functions/sparksql/SIMDJsonFunctions.h:
```diff
@@ -0,0 +1,192 @@
+  // This is a simple validation by checking whether the obtained result is
+  // followed by valid char. Because ondemand parsing we are using ignores json
+  // format validation for characters following the current parsing position.
+  bool isValidEndingCharacter(const char* currentPos) {
```
Are we ensured all valid characters are covered?

mbasmanova replied:
Is this Spark-specific? Do we not need same logic for Presto?

CC: @Yuhta

rui-mo commented on velox/functions/sparksql/SIMDJsonFunctions.h:
```diff
@@ -0,0 +1,192 @@
+      } else {
+        rawResult =
+            ctx.jsonDoc.at_pointer(getFormattedJsonPath(jsonPath).data());
+      }
```
simdjson_result<ondemand::value> rawResult = formattedJsonPath_.has_value() ? one : the other

rui-mo commented on velox/functions/sparksql/tests/JsonFunctionsTest.cpp:
```diff
@@ -0,0 +1,105 @@
+ protected:
+  std::optional<std::string> getJsonObject(
+      std::optional<std::string> json,
+      std::optional<std::string> jsonPath) {
```
const &

rui-mo reviewed (commented):
Can we also update the documentation for get_json_object function?
https://github.com/facebookincubator/velox/blob/main/velox/docs/functions/spark/json.rst#json-functions-1

PHILO-HE commented:
Hi @mbasmanova, could you spare some time to review this pr? Thanks!

mbasmanova commented:
@PHILO-HE CI is red. Would you check?

mbasmanova reviewed (commented):
@PHILO-HE Is this function different from Presto? Curious, what are the differences?

mbasmanova commented on velox/functions/sparksql/Register.cpp:
```diff
@@ -124,7 +124,7 @@ void registerFunctions(const std::string& prefix) {
   // Register size functions
   registerSize(prefix + "size");
 
-  registerFunction<JsonExtractScalarFunction, Varchar, Varchar, Varchar>(
```
Are these any remaining usage of JsonExtractScalarFunction? If not, let's remove it.

mbasmanova replied:
What is JsonExprBenchmark.cpp for? If it is just for benchmarking this function, then since function is no longer used, the benchmark is no longer needed and can be removed.

PHILO-HE replied:
@mbasmanova, I found [JsonExprBenchmark.cpp](https://github.com/facebookincubator/velox/blob/main/velox/functions/prestosql/benchmarks/JsonExprBenchmark.cpp#L34) is used to benchmark a set of functions. Maybe, we can keep it in the source code?

mbasmanova replied:
@PHILO-HE We can keep the benchmark, but there is no point in benchmarking unused JsonExtractScalarFunction, hence, let's remove both the function and benchmark code.

mbasmanova replied:
Alternatively, we can move JsonExtractScalarFunction into the benchmark, assuming it is used to compare folly-based and simd-based implementations.

PHILO-HE replied:
Just removed `JsonExtractScalarFunction` and the related usage in benchmark. Thanks!

mbasmanova commented on velox/functions/sparksql/SIMDJsonFunctions.h:
```diff
@@ -0,0 +1,182 @@
+namespace facebook::velox::functions::sparksql {
+
+template <typename T>
+struct SIMDGetJsonObjectFunction {
```
Is this a copy-paste from somewhere?

mbasmanova commented on velox/functions/sparksql/SIMDJsonFunctions.h:
```diff
@@ -0,0 +1,182 @@
+template <typename T>
+struct SIMDGetJsonObjectFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+  std::optional<std::string> formattedJsonPath_;
```
Make this variable private.

mbasmanova commented on velox/functions/sparksql/SIMDJsonFunctions.h:
```diff
@@ -0,0 +1,182 @@
+
+  // Makes a conversion from spark's json path, e.g., converts
+  // "$.a.b" to "/a/b".
+  FOLLY_ALWAYS_INLINE std::string getFormattedJsonPath(
```
Make this private. Do you really need to inline this method?

mbasmanova commented on velox/functions/sparksql/SIMDJsonFunctions.h:
```diff
@@ -0,0 +1,182 @@
+    }
+  }
+
+  FOLLY_ALWAYS_INLINE simdjson::error_code extractStringResult(
```
private; ditto other places

mbasmanova reviewed (commented):
@PHILO-HE What's the motivation for this PR? Please, update PR description to clarify.

Yuhta commented on velox/functions/sparksql/SIMDJsonFunctions.h:
```diff
@@ -0,0 +1,182 @@
+      if (error) {
+        return false;
+      }
+    } catch (simdjson_error& e) {
```
We should not have any exception here, every error should be represented in the return code

PHILO-HE replied:
Thanks for your comment! 
I found simdjson lib can throw exceptions for some cases.
For example, if we don't keep try/catch here, the below test will fail with an exception thrown:
"_The JSON document has an improper structure: missing or superfluous commas, braces, missing key_"
```c++
EXPECT_EQ(getJsonObject(R"({"hello"-3.5})", "$.hello"), std::nullopt);
```

mbasmanova replied:
@Yuhta Jimmy, I wonder if that's why we are seeing unhandled "Unexpected trailing content in the JSON input" errors in some queries: T175957555 . Since throwing exceptions is expensive (under TRY), I wonder if there is a way to tell simdjson to not throw but rather return error code. Maybe we can open an issue in simdjson GitHub repo and ask about that.

Yuhta replied:
@PHILO-HE @mbasmanova Certain simdjson functions are throwing, you can tell from the signature, but all throwing functions have a corresponding version that is returning error code.  So we need to find out which function is throwing and replace them, then we can remove the catch here.

PHILO-HE commented:
> @PHILO-HE What's the motivation for this PR? Please, update PR description to clarify.

Hi @mbasmanova, I just updated PR description and fixed the comments. Thanks for your review!

PHILO-HE commented:
@mbasmanova, could you take a review further? Thanks!

mbasmanova reviewed (commented):
@PHILO-HE 

> Add simdjson based get_json_object Spark function

The PR title suggests that this PR adds a new Spark function. However, the description suggests that the changes are to optimize an existing function not to add a new function. Which is correct? It would be nice to align PR title, PR description, and code changes.

mbasmanova commented on velox/docs/functions/spark/json.rst:
```diff
@@ -22,6 +22,8 @@ JSON Functions
 .. spark:function:: get_json_object(json, path) -> varchar
 
-    Extracts a json object from path::
+    Extracts a json object from ``path``. Returns NULL if it finds json string
```
It would be nice to document what values for 'path' are supported? Maybe provide a link to some existing documentation for path?

PHILO-HE replied:
@mbasmanova, I didn't find official document for valid json path. Just documented the pattern and gave some examples.

mbasmanova commented on velox/docs/functions/spark/json.rst:
```diff
@@ -22,6 +22,8 @@ JSON Functions
-        SELECT get_json_object('{"a":"b"}', '$.a'); -- b
\ No newline at end of file
+        SELECT get_json_object('{"a":"b"}', '$.a'); -- 'b'
+        SELECT get_json_object('{"a":{"b":"c"}}', '$.a'); -- '{"b":"c"}'
```
Does this function take an argument of type VARCHAR and return result of type JSON? (this is what signature on L23 suggests)? Or, does this function return VARCHAR result?

PHILO-HE replied:
All inputs are VARCHAR and output is also VARCHAR. Just clarified in the doc. Thanks!

mbasmanova commented on velox/functions/sparksql/SIMDJsonFunctions.h:
```diff
@@ -0,0 +1,178 @@
+      const arg_type<Varchar>& jsonPath) {
+    char formattedJsonPath[jsonPath.size() + 1];
+    int j = 0;
+    for (int i = 0; i < jsonPath.size(); i++) {
```
This logic is pretty important as it effectively defines what paths are supported. It would be nice to extract it into a separate header with clear documentation and dedicated unit tests.

For example, looks like we allow $ to appear anywhere within the path. Is this correct? Can you provide an example of a valid 'path' where $ appears somewhere in the middle?

PHILO-HE replied:
>  looks like we allow $ to appear anywhere within the path. Is this correct?

Just found this code is wrong. `$` should be in the initial position and should not occur after this, which is verified with Spark query. I just updated this code and added some test.

PHILO-HE commented on velox/docs/functions/spark/json.rst:
```diff
@@ -20,8 +20,14 @@ such as ``[1,2,3]``. More detailed grammar can be found in
 
-    Extracts a json object from path::
+    Returns a json object, represented by VARCHAR, from ``jsonString`` by searching ``path``.
+    Valid ``path`` should contain "[index]", "[field]" or ".field" to define a JSON path.
```
Yes, Spark requires `$` is the initial character in `path`. Just updated the doc.
Here is spark code, which starts the parsing from "$".
https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala#L94

PHILO-HE commented:
Hi @mbasmanova, could you review this pr further?

As discussed before, `getJsonPointerPath` is important since it converts json path to json pointer path used for searching, e.g, from "$.a.b[0]" to "/a/b/0". I recently found such conversion has already been implemented in simdjson lib, but only supported from 3.8.0 (the version is 3.2.0 in velox by now). So I copied those code from simdjson lib for path conversion.

mbasmanova commented:
> I recently found such conversion has already been implemented in simdjson lib, but only supported from 3.8.0 (the version is 3.2.0 in velox by now). So I copied those code from simdjson lib for path conversion.

Let's upgrade to simdjson 3.8.0.

PHILO-HE commented:
> > I recently found such conversion has already been implemented in simdjson lib, but only supported from 3.8.0 (the version is 3.2.0 in velox by now). So I copied those code from simdjson lib for path conversion.
> 
> Let's upgrade to simdjson 3.8.0.

Hi @mbasmanova, as simdjson has been upgraded, I just updated this pr to use its new API. Could you please review again? Thanks!

PHILO-HE commented:
@mbasmanova, could you review this pr again? Thanks!

mbasmanova commented on velox/functions/sparksql/SIMDJsonFunctions.h:
```diff
@@ -0,0 +1,195 @@
+
+#include "velox/functions/prestosql/SIMDJsonFunctions.h"
+
+using namespace simdjson;
```
no 'using namespace' in header files, please

mbasmanova commented on velox/functions/sparksql/SIMDJsonFunctions.h:
```diff
@@ -0,0 +1,195 @@
+namespace facebook::velox::functions::sparksql {
+
+template <typename T>
+struct SIMDGetJsonObjectFunction {
```
Let's remove 'SIMD' prefix. Presto functions have it now, because, originally there were 2 sets of functions, but we should rename these as well.

mbasmanova commented on velox/functions/sparksql/SIMDJsonFunctions.h:
```diff
@@ -0,0 +1,195 @@
+      }
+      pairEnd = result.find("]", pairBegin);
+      if (pairEnd == std::string::npos || result[pairEnd - 1] != '\'') {
+        return "-1";
```
This seems a bit hacky. Are you relying on simdjson to reject this path later? It would be cleaner to raise an exception here and provide a clear error message.

This function seems to be working around a limitation on simdjson. Have you already asked in simdjson project if that limitation can be removed?  Anyway, please, update PR description to explain this limitation and how you are working around it.

PHILO-HE replied:
> Are you relying on simdjson to reject this path later? It would be cleaner to raise an exception here and provide a clear error message.

@mbasmanova, yes, we depend on simdjson to return error code for illegal path, then this function will return NULL result at last, instead of throwing exception.

> Have you already asked in simdjson project if that limitation can be removed?

I just left a comment in simdjson community to discuss this limitation:
https://github.com/simdjson/simdjson/issues/2070#issuecomment-2079688880

mbasmanova replied:
@PHILO-HE 

The API that returns NULL when json or path are invalid is challenging. Throwing an error would be better.

Note that simdjson has limited support for JSONPath, hence, it rejects both invalid and valid-but-not-supported paths. If Spark supports full JSONPath spec or a wider subset of the spec than simdjson, then this implementation will produce incorrect results.

Hence, some questions.

- What subset of JSONPath is supported in Spark?
- Is there any particular reason to not use SIMDJsonExtract[Scalar]Function directly or implement this in a similar way by leveraging velox/functions/prestosql/json/SIMDJsonExtractor.h ?
- Does Spark validate JSON document fully and return null if it is not valid? simdjson doesn't do that and may succeed at extracting the value from an invalid document. This will then cause Gluten to produce wrong results.

PHILO-HE replied:
> @PHILO-HE
> 
> The API that returns NULL when json or path are invalid is challenging. Throwing an error would be better.
> 
> Note that simdjson has limited support for JSONPath, hence, it rejects both invalid and valid-but-not-supported paths. If Spark supports full JSONPath spec or a wider subset of the spec than simdjson, then this implementation will produce incorrect results.
> 
> Hence, some questions.
> 
> * What subset of JSONPath is supported in Spark?

@mbasmanova, thanks for your comment!
Spark has its own code to parse Json path. I have addressed all inconsistency found in Spark tests. For example, ['name']['id'] is supported by Spark, but not by Simdjson. This pr fixed this inconsistency by just pre-processing the path. So I feel it's better to return NULL instead of throwing exception for invalid path.

> * Is there any particular reason to not use SIMDJsonExtract[Scalar]Function directly or implement this in a similar way by leveraging velox/functions/prestosql/json/SIMDJsonExtractor.h ?

The proposed pr has some handlings to align with Spark, e.g., check the validity of JSON document. And with this proposed patch, all Spark test cases can pass.

> * Does Spark validate JSON document fully and return null if it is not valid? simdjson doesn't do that and may succeed at extracting the value from an invalid document. This will then cause Gluten to produce wrong results.

Yes, Spark fully validates JSON document. Simdjson only checks the validity before the position of extracted result. That's why this pr checked the characters after this position. Though it's not a full validation, it can meet our Spark users' requirement.

mbasmanova commented on velox/functions/sparksql/SIMDJsonFunctions.h:
```diff
@@ -0,0 +1,195 @@
+  // Spark's json path requires field name surrounded by single quotes if it is
+  // specified in "[]". But simdjson lib requires not. This method just removes
+  // such single quotes, e.g., converts "['a']['b']" to "[a][b]".
+  FOLLY_ALWAYS_INLINE std::string removeSingleQuotes(
```
no need for FOLLY_ALWAYS_INLINE here; please, remove

mbasmanova commented on velox/functions/sparksql/SIMDJsonFunctions.h:
```diff
@@ -0,0 +1,195 @@
+      const arg_type<Varchar>& json,
+      const arg_type<Varchar>& jsonPath) {
+    // Spark requires the first char in jsonPath is '$'.
+    if (jsonPath.size() < 1 || jsonPath.data()[0] != '$') {
```
this check is repeated in L36; perhaps, add a small helper function to avoid copy-paste

mbasmanova commented on velox/functions/sparksql/SIMDJsonFunctions.h:
```diff
@@ -0,0 +1,195 @@
+        ? ctx.jsonDoc.at_path(jsonPath_.value().data())
+        : ctx.jsonDoc.at_path(
+              removeSingleQuotes(
+                  std::string_view(jsonPath.data() + 1, jsonPath.size() - 1))
```
> std::string_view(jsonPath.data() + 1, jsonPath.size() - 1)

this core repeats; better change the signature of removeSingleQuotes to take path as is:

```
removeSingleQuotes(path)
```

mbasmanova commented on velox/functions/sparksql/SIMDJsonFunctions.h:
```diff
@@ -0,0 +1,195 @@
+    if (rawResult.error()) {
+      return false;
+    }
+    auto error = extractStringResult(rawResult, result);
```
error variable is not needed

```
if (!extractStringResult(...)) {
   return false;
}
```

mbasmanova commented on velox/functions/sparksql/SIMDJsonFunctions.h:
```diff
@@ -0,0 +1,195 @@
+    // These chars can be prior to a valid ending char.
+    if (endingChar == ' ' || endingChar == '\r' || endingChar == '\n' ||
+        endingChar == '\t') {
+      return isValidEndingCharacter(currentPos++);
```
What if currentPos++ is past the end of the doc?

PHILO-HE replied:
I note json doc is padded with '\0' character. So it is safe to do recursively check. Just added a comment to clarify this.

mbasmanova commented on velox/functions/sparksql/SIMDJsonFunctions.h:
```diff
@@ -0,0 +1,195 @@
+    return result;
+  }
+
+  FOLLY_ALWAYS_INLINE simdjson::error_code extractStringResult(
```
FOLLY_ALWAYS_INLINE not needed; please, remove

mbasmanova commented on velox/functions/sparksql/SIMDJsonFunctions.h:
```diff
@@ -0,0 +1,195 @@
+            uint64_t numberResult;
+            error = rawResult.get_uint64().get(numberResult);
+            if (!error) {
+              ss << numberResult;
```
Why do we need to do this conversion via std::stringstream ?

PHILO-HE replied:
@mbasmanova, I feel handy to use std::stringstream here. Do you have any suggestion? Thanks!

mbasmanova commented on velox/functions/sparksql/SIMDJsonFunctions.h:
```diff
@@ -0,0 +1,195 @@
+  FOLLY_ALWAYS_INLINE simdjson::error_code extractStringResult(
+      simdjson_result<ondemand::value> rawResult,
+      out_type<Varchar>& result) {
+    simdjson::error_code error;
```
this variable is not needed; this function can simply return a boolean

mbasmanova commented on velox/functions/sparksql/SIMDJsonFunctions.h:
```diff
@@ -0,0 +1,195 @@
+      // can make simdjson's internal parsing position moved and then we
+      // can check the validity of ending character.
+      case ondemand::json_type::number: {
+        switch (rawResult.get_number_type()) {
```
I wonder if this code repeats somewhere else. Would you check?

PHILO-HE replied:
Just checked again. I didn't find the repeat code anywhere else.

mbasmanova commented on velox/functions/sparksql/tests/JsonFunctionsTest.cpp:
```diff
@@ -0,0 +1,119 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "velox/functions/prestosql/types/JsonType.h"
```
Why is this include? We shouldn't use Presto types in Spark code.

mbasmanova commented on velox/functions/sparksql/tests/JsonFunctionsTest.cpp:
```diff
@@ -0,0 +1,119 @@
+          "$[1].other[1]"),
+      "v2");
+
+  // Field not found.
```
This test is rather long. Consider extracting invalid cases into a separate test method.

mbasmanova commented on velox/functions/sparksql/tests/JsonFunctionsTest.cpp:
```diff
@@ -0,0 +1,119 @@
+class JsonFunctionTest : public SparkFunctionBaseTest {
+ protected:
+  std::optional<std::string> getJsonObject(
+      const std::optional<std::string>& json,
```
No need to use optional here as this method doesn't allow unset inputs.

mbasmanova reviewed (commented):
@PHILO-HE Some comments. Please, rebase.

PHILO-HE commented on velox/functions/prestosql/benchmarks/JsonExprBenchmark.cpp:
```diff
@@ -189,8 +189,6 @@ class JsonBenchmark : public velox::functions::test::FunctionBenchmarkBase {
         {"folly_json_array_length"});
     registerFunction<SIMDJsonArrayLengthFunction, int64_t, Json>(
         {"simd_json_array_length"});
-    registerFunction<JsonExtractScalarFunction, Varchar, Json, Varchar>(
```
@mbasmanova, I note you have moved `JsonExtractScalarFunction` to this file. Do I need to put the deleted code back for this file?

PHILO-HE replied:
@mbasmanova, just reverted. Thanks!

mbasmanova commented on velox/functions/sparksql/JsonFunctions.h:
```diff
@@ -0,0 +1,201 @@
+      }
+      pairEnd = result.find("]", pairBegin);
+      if (pairEnd == std::string::npos || result[pairEnd - 1] != '\'') {
+        return "-1";
```
Let's not pass knowngly invalid path to simdjson, but instead cut processing short and return NULL.

mbasmanova commented:
I found some info about get_json_path implementation in Spark.

https://issues.apache.org/jira/browse/SPARK-37857

```
It seems the first implementation of Spark get_json_object does replicate Hive behaviour (https://github.com/apache/spark/pull/7901)

In hive documentation (https://cwiki.apache.org/confluence/display/hive/languagemanual+udf) it is stated that recursive decent (i.e ".." notation) is not supported
```

A limited version of JSONPath is supported:

- $ : Root object
- . : Child operator
- [] : Subscript operator for array
- `*` : Wildcard for []

Syntax not supported that's worth noticing:

- : Zero length string as key
- .. : Recursive descent
- @ : Current object/element
- () : Script expression
- ?() : Filter (script) expression.
- [,] : Union operator
- [start:end.step] : array slice operator
